[RF DOCS] Action Mailbox Guides (#1)
bhumi1102/rails: guides/source/action_mailbox_basics.md

What is Action Mailbox?
-----------------------

Action Mailbox routes incoming emails to controller-like mailboxes for processing in your Rails application. Action Mailbox is for receiving inbound email, while [Action Mailer](https://guides.rubyonrails.org/action_mailer_basics.html) is for *sending* emails from your Rails application.

The inbound emails are routed asynchronously using Active Job to one or several dedicated mailboxes. These emails are turned into `InboundEmail` records using Active Record, which are capable of interacting directly with the rest of your domain model.

> `InboundEmail` records also provide lifecycle tracking, storage of the original email via Active Storage, and responsible data handling with on-by-default incineration.

Reviewer: "Incineration" is a new word/concept, maybe we could also link to the incineration section within the guide?

> Action Mailbox ships with ingresses for external email providers such as Mailgun, Mandrill, Postmark, and SendGrid. You can also handle inbound mails directly via the built-in Exim, Postfix, and Qmail ingresses.

Reviewer 2: Suggestion: Do you think it's worth adding links to the external email providers?

Reviewer 3: My vote would be: not necessary.

Reviewer: Maybe not externally, but perhaps linking to the individual subsections within the guide for easier nav?

Author: Agree no external links. Links to their configuration docs are in the individual subsections.

> Action Mailbox ships with ingresses for external email providers such as Mailgun, Mandrill, Postmark, and SendGrid. You can also handle inbound mails directly via the built-in Exim, Postfix, and Qmail ingresses.

Reviewer 3: Can we add just a little bit of context about what 'ingresses' are? Devs who have done this before will understand, but I am thinking of the (new) dev implementing this for the first time. Something like: "Action Mailbox ships with ingresses which enable your application to receive emails from external email providers and such as Mailgun, etc..... "

## Setup

Action Mailbox has a few moving parts. First you'll run the Action Mailbox installer. Then configure a chosen ingress for handling incoming email. You're then ready to add Action Mailbox routing, create mailboxes and start processing incoming emails.

To start let's install Action Mailbox:

> 1. `bin/rails action_mailbox:install` - this will create an `application_mailbox.rb` file and copy over migrations.

Reviewer 3: I think we might want to keep the $. We have a tip in the getting started guide: "The examples below use $ to represent your terminal prompt in a UNIX-like OS, though it may have been customized to appear differently. If you are using Windows, your prompt will look something like C:\source_code>."

> 1. `bin/rails action_mailbox:install` - this will create an `application_mailbox.rb` file and copy over migrations.
> 2. `bin/rails db:migrate` - this will run the Action Mailbox and Active Storage migrations.

Reviewer: One thing that just occurred to me, related to what Amanda mentioned above: we might want to try to keep commands split into their own blocks, that makes it easier to copy manually, but also gives readers a free "copy" button to click:

<img width="653" alt="Screenshot 2024-02-05 at 11 59 38" src="https://github.com/bhumi1102/rails/assets/26328/a1dd88f0-c73f-4cf5-9ae2-2bca6554f3b6">

We could use comments to connect things: (they aren't copied with that button)

````suggestion
```bash
# Creates an `application_mailbox.rb` file, copies over migrations.
$ bin/rails action_mailbox:install
```

```bash
# Runs the Action Mailbox and Active Storage migrations.
$ bin/rails db:migrate
```
````

The output would look something like this:

<img width="663" alt="Screenshot 2024-02-05 at 12 04 52" src="https://github.com/bhumi1102/rails/assets/26328/092c8d23-55fd-42a7-a405-f59e30153e61">

Or with paragraphs:

````suggestion
```bash
$ bin/rails action_mailbox:install
```

This will create an `application_mailbox.rb` file and copy over migrations.

```bash
$ bin/rails db:migrate
```

This will run the Action Mailbox and Active Storage migrations.
````

_(note that I merged the last sentence with the subsequent paragraph in the screenshot below)_

<img width="686" alt="Screenshot 2024-02-05 at 12 06 30" src="https://github.com/bhumi1102/rails/assets/26328/5600e636-5375-459f-9c27-e9412eace8ba">

Author: Good point about the 'copy' button! I like the one with paragraphs a bit better, so going with that suggestion.

> The Action Mailbox table `action_mailbox_inbound_emails` stores incoming message and their processing 'status'.

Reviewer 2: ```suggestion
The Action Mailbox table `action_mailbox_inbound_emails` stores incoming messages and their processing 'status'.
```

> At this point you can start your rails server and check out `http://localhost:3000/rails/conductor/action_mailbox/inbound_emails`.

Reviewer: ~Should we expand on this to at least mention what you can do there?~
~Just to give some extra context of how that can be used and useful during dev.~

---

I was going to suggest the above, but realized there's a section below about local dev that mentions it and explains better. Maybe we can just link to it from here?

The next step is to configure an ingress in your Rails application to specify how incoming emails should be received.

## Ingress Configuration

> Configuring ingress involves setting up credentials and endpoint information for the chosen email service. Here are the steps for each of the supported ingress.

Reviewer 3: Would it be "each of the supported ingress**es**"? I understand that "configuring ingress" is used when discussing the concept, but not sure if that extends to the plural when you say "each of the ingress"?

What is Action Mailbox?
-----------------------

> Action Mailbox routes incoming emails to controller-like mailboxes for processing in your Rails application. Action Mailbox is for receiving email, while [Action Mailer](https://guides.rubyonrails.org/action_mailer_basics.html) is for *sending* them.
> 
> The inbound emails are routed asynchronously using [Active Job](https://guides.rubyonrails.org/active_job_basics.html) to one or several dedicated mailboxes. These emails are turned into [`InboundEmail`](https://api.rubyonrails.org/classes/ActionMailbox/InboundEmail.html) records using [Active Record](https://guides.rubyonrails.org/active_record_basics.html), which are capable of interacting directly with the rest of your domain model.
> 
> `InboundEmail` records also provide lifecycle tracking, storage of the original email via [Active Storage](https://guides.rubyonrails.org/active_storage_overview.html), and responsible data handling with on-by-default [incineration](#incineration-of-inboundemails).

Reviewer: Note for future linking: it's possible to link to another guide by simply pointing to the HTML name, it's just slightly easier to type, but also allows you to generate everything locally and navigate around. (I don't think it works for api, but for linking between guides it can be handy.)

```suggestion
Action Mailbox routes incoming emails to controller-like mailboxes for processing in your Rails application. Action Mailbox is for receiving email, while [Action Mailer](action_mailer_basics.html) is for *sending* them.

The inbound emails are routed asynchronously using [Active Job](active_job_basics.html) to one or several dedicated mailboxes. These emails are turned into [`InboundEmail`](https://api.rubyonrails.org/classes/ActionMailbox/InboundEmail.html) records using [Active Record](active_record_basics.html), which are capable of interacting directly with the rest of your domain model.

`InboundEmail` records also provide lifecycle tracking, storage of the original email via [Active Storage](active_storage_overview.html), and responsible data handling with on-by-default [incineration](#incineration-of-inboundemails).
```

Action Mailbox ships with ingresses which enable your application to receive emails from external email providers such as Mailgun, Mandrill, Postmark, and SendGrid. You can also handle inbound emails directly via the built-in Exim, Postfix, and Qmail ingresses.

## Setup

> Action Mailbox has a few moving parts. First you'll run the Action Mailbox installer. Then configure a chosen ingress for handling incoming email. You're then ready to add Action Mailbox routing, create mailboxes and start processing incoming emails.

Reviewer: Noticed we're mentioning `Action Mailbox` 3 times in this intro paragraph, maybe we can slightly simplify?

```suggestion
Action Mailbox has a few moving parts. First you'll run the installer. Then configure a chosen ingress for handling incoming email. You're then ready to add Action Mailbox routing, create mailboxes and start processing incoming emails.
```

or maybe:

```suggestion
Action Mailbox has a few moving parts: first you'll run the installer. Then configure a chosen ingress for handling incoming email. You're then ready to add Action Mailbox routing, create mailboxes and start processing incoming emails.
```

NOTE: When configuring your SendGrid Inbound Parse webhook, be sure to check the box labeled **“Post the raw, full MIME message.”** Action Mailbox needs the raw MIME message to work.

## Processing Incoming Email

Processing incoming emails usually entails using the email content to create models, update views, queue background work, etc. in your Rails application.

Before you can start processing incoming emails, you'll need to setup Action Mailbox routing and create mailboxes.

> ### Configure routing

Reviewer 4: For the guide headings we try to:
> capitalize all words except for prepositions, conjunctions, internal articles, and forms of the verb "to be"

https://guides.rubyonrails.org/ruby_on_rails_guides_guidelines.html#headings
```suggestion
### Configure Routing
```

Author: Thanks @p8

> Routes for Action Mailbox are added to the `application_mailbox.rb` file. The regular expression matches the incoming email's `to` field.

Reviewer: Maybe we can expand a bit on what "routing" does / why it's needed? It might not be as clear to new people.

```suggestion
After an incoming email is received via the configured ingress, it needs to be forwarded to a mailbox for actual processing by your application. Routing in Action Mailbox is where we define which emails go to which mailboxes for processing. They are added to the `application_mailbox.rb` file. The regular expression matches the incoming email's `to` field:
```

Another idea, that makes the connection with the app routing system:

```suggestion
After an incoming email is received via the configured ingress, it needs to be forwarded to a mailbox for actual processing by your application. Much like the [Rails router that dispatches URLs to controllers](https://guides.rubyonrails.org/routing.html), routing in Action Mailbox defines which emails go to which mailboxes for processing. They are added to the `application_mailbox.rb` file. The regular expression matches the incoming email's `to` field:
```

These are just suggestions to give an idea, not necessarily final copy, so feel free to tweak as you see fit.

NOTE: When configuring your SendGrid Inbound Parse webhook, be sure to check the box labeled **“Post the raw, full MIME message.”** Action Mailbox needs the raw MIME message to work.

## Processing Incoming Email

Processing incoming emails usually entails using the email content to create models, update views, queue background work, etc. in your Rails application.

Before you can start processing incoming emails, you'll need to setup Action Mailbox routing and create mailboxes.

### Configure Routing

> After an incoming email is received via the configured ingress, it needs to be forwarded to a mailbox for actual processing by your application. Much like the [Rails router](https://guides.rubyonrails.org/routing.html) that dispatches URLs to controllers, routing in Action Mailbox defines which emails go to which mailboxes for processing. Routes are added to the `application_mailbox.rb` file using regular expression:

Reviewer: My bad, I think I previously referenced the full guide URL when I sent feedback 😅, but we can / should link to the file only for the same reasons I mentioned on another comment. (easier to test locally, etc.)

```suggestion
After an incoming email is received via the configured ingress, it needs to be forwarded to a mailbox for actual processing by your application. Much like the [Rails router](routing.html) that dispatches URLs to controllers, routing in Action Mailbox defines which emails go to which mailboxes for processing. Routes are added to the `application_mailbox.rb` file using regular expression:
```

Author: No worries. I *thought* I caught that but fixed now!

end
```

> For example, the above will match any email sent to `save@` to a 'forwards' mailbox. Which we need to create.

Reviewer 2: ```suggestion
For example, the above will match any email sent to `save@` to a 'forwards' mailbox, which we will need to create.
```

> ### Create a mailbox

Reviewer 4: ```suggestion
### Create a Mailbox
```

```bash
# Generate new mailbox
$ bin/rails generate mailbox forwards
```

> This creates a `ForwardsMailbox` class with a `process` method.

Reviewer: Perhaps we can mention where those mailboxes are located? e.g.:
 
```suggestion
This creates `app/mailboxes/forwards_mailbox.rb`, with a `ForwardsMailbox` class and a `process` method.
```

> ### Process email

Reviewer 2: Love this section 💜

> ### Process email

Reviewer 4: ```suggestion
### Process Email
```

#<Mail::Message:33780, Multipart: false, Headers: <Date: Wed, 31 Jan 2024 22:18:40 -0600>, <From: someone@hey.com>, <To: save@example.com>, <Message-ID: <65bb1ba066830_50303a70397e@Bhumis-MacBook-Pro.local.mail>>, <In-Reply-To: >, <Subject: Hello Action Mailbox>, <Mime-Version: 1.0>, <Content-Type: text/plain; charset=UTF-8>, <Content-Transfer-Encoding: 7bit>, <x-original-to: >>
$ mail.to
["save@example.com"]
$ mail.from
["someone@hey.com"]
$ mail.date
Wed, 31 Jan 2024 22:18:40 -0600
$ mail.subject
"Hello Action Mailbox"
$ mail.body
#<Mail::Body:0x00007fc74cbf46c0 @boundary=nil, @preamble=nil, @epilogue=nil, @charset="US-ASCII", @part_sort_order=["text/plain", "text/enriched", "text/html", "multipart/alternative"], @parts=[], @raw_source="This is the body of the email message.", @ascii_only=true, @encoding="7bit">
$ mail.body.decoded
"This is the body of the email message."
```
### Inbound Email Status

Once an email has been routed to the matching Mailbox and processed, Action Mailbox updates the email status stored in `action_mailbox_inbound_emails` table with one of the following values:

- Pending: Just received by one of the ingress controllers and scheduled for routing.

> - Processing: During active processing, while a specific mailbox is running its process method.

Reviewer 2: Should we reference it like #process like we sis for Failed on line 326?

```suggestion
- Processing: During active processing, while a specific mailbox is running its #process method.
```

Reviewer: Either that, or perhaps wrap both instances to highlight as code:

```suggestion
- Processing: During active processing, while a specific mailbox is running its `process` method.
```

Author: Good call. I like `process`

- Delivered: Successfully processed by the specific mailbox.
- Failed: An exception was raised during the specific mailbox’s execution of the #process method.
- Bounced: Rejected processing by the specific mailbox and bounced to sender.

> If the email is marked either `delivered`, `failed`, or `bounced` it's considered 'processed' and marked for incineration.

Reviewer: Might be good to link to the incineration section here as well.

[`Mail`]: https://github.com/mikel/mail

## Example

Here is an example of an Action Mailbox that processes emails to create 'forwards' for the user's project.

> `before_processing` callback is used to ensure that certain conditions are met before `process` is called. In this case, the email is bounced using `bounced_with` if the 'forwarder' has no projects. The 'forwarder' is a `User` with the same email as `mail.from`.

Reviewer 2: ```suggestion
The `before_processing` callback is used to ensure that certain conditions are met before `process` is called. In this case, the email is bounced using `bounced_with` if the 'forwarder' has no projects. The 'forwarder' is a `User` with the same email as `mail.from`.
```

Reviewer: Since this mentions a callback, and there's before/after/around processing, maybe we can add a note about their existence around here?

There's probably no need for a full section on callbacks (although we could consider it too), but mentioning they exist in the guide might be nice.

If the 'forwarder' does have at least one project, the `record_forward` method creates an Active Record model in the application using the email data `mail.subject` and `mail.content`. Otherwise it sends an email, using Action Mailer, requesting the 'forwarder' to choose a project.

> 

Reviewer: Nit: double blank lines ✂️

end
```

> The regular expression matches the incoming email's `to` field. For example, the above will match any email sent to `save@` to a 'forwards' mailbox, which we will need to create.

Reviewer: Nice :+1:

I just remembered there's a few other ways you can add "routing" that aren't regex based, maybe we could point people to the API to learn more about them? https://api.rubyonrails.org/classes/ActionMailbox/Base.html

Author: Totally. Good call!

$ mail.to
["save@example.com"]
$ mail.from
["someone@hey.com"]
$ mail.date
Wed, 31 Jan 2024 22:18:40 -0600
$ mail.subject
"Hello Action Mailbox"
$ mail.body
#<Mail::Body:0x00007fc74cbf46c0 @boundary=nil, @preamble=nil, @epilogue=nil, @charset="US-ASCII", @part_sort_order=["text/plain", "text/enriched", "text/html", "multipart/alternative"], @parts=[], @raw_source="This is the body of the email message.", @ascii_only=true, @encoding="7bit">
$ mail.body.decoded
"This is the body of the email message."
```
### Inbound Email Status

Once an email has been routed to a matching mailbox and processed, Action Mailbox updates the email status stored in `action_mailbox_inbound_emails` table with one of the following values:

- Pending: Just received by one of the ingress controllers and scheduled for routing.
- Processing: During active processing, while a specific mailbox is running its `process` method.
- Delivered: Successfully processed by the specific mailbox.

> - Failed: An exception was raised during the specific mailbox’s execution of the #process method.

Reviewer: For consistency with the other mention above.

```suggestion
- Failed: An exception was raised during the specific mailbox’s execution of the `process` method.
```

> - Pending: Just received by one of the ingress controllers and scheduled for routing.
> - Processing: During active processing, while a specific mailbox is running its `process` method.
> - Delivered: Successfully processed by the specific mailbox.
> - Failed: An exception was raised during the specific mailbox’s execution of the #process method.
> - Bounced: Rejected processing by the specific mailbox and bounced to sender.

Reviewer: Perhaps we should use the actual statuses as code here, like in the paragraph below, to avoid potential confusion, and make a better connection between them?

e.g.:

```suggestion
- `pending`: Just received by one of the ingress controllers and scheduled for routing.
- `processing`: During active processing, while a specific mailbox is running its `process` method.
- `delivered`: Successfully processed by the specific mailbox.
- `failed`: An exception was raised during the specific mailbox’s execution of the #process method.
- `bounced`: Rejected processing by the specific mailbox and bounced to sender.
```

Author: Yeah I like that better!

If the email is marked either `delivered`, `failed`, or `bounced` it's considered 'processed' and marked for [incineration](#incineration-of-inboundemails).

## Example

Here is an example of an Action Mailbox that processes emails to create 'forwards' for the user's project.

> The `before_processing` callback is used to ensure that certain conditions are met before `process` method is called. In this case, `before_processing` checks that the user has at least on project. Other supported [Action Mailbox callbacks](https://api.rubyonrails.org/classes/ActionMailbox/Callbacks.html) are `after_processing` and `around_processing`.

Reviewer: ```suggestion
The `before_processing` callback is used to ensure that certain conditions are met before `process` method is called. In this case, `before_processing` checks that the user has at least one project. Other supported [Action Mailbox callbacks](https://api.rubyonrails.org/classes/ActionMailbox/Callbacks.html) are `after_processing` and `around_processing`.
```

```

Please refer to the [ActionMailbox::TestHelper API](https://api.rubyonrails.org/classes/ActionMailbox/TestHelper.html) for further test helper methods.

## Incineration of InboundEmails

> By default, an [`InboundEmail`][] that has been processed will be incinerated after 30 days. The InboundEmail is considered as processed when its status changes to `delivered`, `failed` or `bounced`.

Reviewer: Maybe we should link to InboundEmail on its first appearance under "What is Action Mailbox?" instead?

> The actual incineration is done via the `IncinerationJob` that's scheduled to run after [`config.action_mailbox.incinerate_after`][] time. This value is set to `30.days` by default, but you can change it in your production.rb configuration. (Note that this far-future incineration scheduling relies on your job queue being able to hold jobs for that long.)

Reviewer: May be worth linking to the `IncinerationJob` as well.
